include/boost/windows_sspi/detail/context_impl.hpp

```diff
@@ -11,150 +11,43 @@
-                             static_cast<boost::winapi::DWORD_>(ca.size()),
-                             0,
-                             buffer.data(),
-                             &size,
-                             nullptr,
-                             nullptr)) {
-      ec.assign(boost::winapi::GetLastError(), boost::system::system_category());
-      return;
-    }
-
-    cert_context_type cert{CertCreateCertificateContext(X509_ASN_ENCODING, buffer.data(), size),
-      CertFreeCertificateContext};
-    if (!cert) {
-      ec.assign(boost::winapi::GetLastError(), boost::system::system_category());
-      return;
-    }
-
+  void add_certificate_authority(const net::const_buffer& ca) {
+    cert_context cert{pem_to_cert_context(ca), &CertFreeCertificateContext};
     if(!CertAddCertificateContextToStore(m_cert_store,
                                          cert.get(),
                                          CERT_STORE_ADD_ALWAYS,
                                          nullptr)) {
-      ec.assign(boost::winapi::GetLastError(), boost::system::system_category());
-      return;
+      throw boost::system::system_error(boost::winapi::GetLastError(), boost::system::system_category());
     }
   }
 
-  void load_verify_file(const std::string& filename, boost::system::error_code& ec) {
-    using file_handle_type = std::unique_ptr<std::remove_pointer<boost::winapi::HANDLE_>::type,
-                                             decltype(&boost::winapi::CloseHandle)>;
-
-    file_handle_type handle{boost::winapi::CreateFile(filename.c_str(),
-                                                      boost::winapi::GENERIC_READ_,
-                                                      boost::winapi::FILE_SHARE_READ_,
-                                                      nullptr,
-                                                      boost::winapi::OPEN_EXISTING_,
-                                                      boost::winapi::FILE_ATTRIBUTE_NORMAL_,
-                                                      nullptr),
-      boost::winapi::CloseHandle};
-    if (handle.get() == boost::winapi::INVALID_HANDLE_VALUE_) {
-      ec.assign(boost::winapi::GetLastError(), boost::system::system_category());
-      return;
-    }
-
-    boost::winapi::LARGE_INTEGER_ size;
-    if(!boost::winapi::GetFileSizeEx(handle.get(), &size)) {
-      ec.assign(boost::winapi::GetLastError(), boost::system::system_category());
-      return;
-    }
-
-    std::vector<char> buffer(static_cast<std::size_t>(size.QuadPart));
-    boost::winapi::DWORD_ read;
-    if(!boost::winapi::ReadFile(handle.get(), buffer.data(), static_cast<boost::winapi::DWORD_>(buffer.size()), &read, nullptr)) {
-      ec.assign(boost::winapi::GetLastError(), boost::system::system_category());
-      return;
-    }
-
-    add_certificate_authority(net::buffer(buffer), ec);
+  void load_verify_file(const std::string& filename) {
```

Comment:
Why `string` instead of `string_view`?

Reply:
Again, this is the interface in `boost::asio::ssl` so I'm trying to use the same.

Reply:
asio uses `std::string` because it supports C++03 and the SSL code is ancient. Since you require C++11 (I hope) and Boost, you can use `boost::string_view` instead, which is a better choice.

Reply:
I've sort of decided to try and make this a "drop in replacement" for `boost::asio::ssl`, so I prefer to just copy the existing interface.

My goal is to be able to take existing `boost::asio::ssl` code and only change a namespace and everything should work as before, so I prefer not to change this.

Of course the caller could most likely call the function the same way whether it's a `std::string` or a `boost::string_view`, but I think it's preferable that the functions look the same.

Reply:
My 2c here is that if the underlying C API is going to require a null-terminated string and the parameter type to the wrapper is a `string_view`, then the wrapper will need to create a `std::string` (or similar copy) of the text being passed. In this case, having `std::string` as the interface type is no worse, and often better than using a `string_view` - at least in terms of byte copies and possible memory allocation.

If the underlying API is either: 
a) going to make a copy anyway, as in the case of an async initiating function, or
b) or act on the sequence of bytes and return immediately without needing a null termination,

then a `string_view` can avoid the creation of an un-necessary string at the client site.

Having said all this, **as of c++20**, `std::string` becomes `constexpr` if used immediately and discarded (at least according to the standard) so the creation of the `std::string` at the client site can actually be optimised away.

Reply:
@madmongo1 Thanks a lot for your input.

Since the argument is a path to a file in the filesystem, I think the correct type to use would actually be `boost::filesystem::path` or `std::filesystem::path`. That should also handle unicode characters in the filename, but that would of course add a dependency on either `boost::filesystem` or C++17.

As I mentioned, I originally wanted this to be a "drop in replacement" for `asio::ssl`, so I should imply use the same signature as the existing function from `asio::ssl`.

Thinking a bit more about that and after a very valuable talk with @vinniefalco on the Slack channel, I've come to the conclusion that it doesn't make much sense to aim for API compatibility with `asio::ssl` anyway, so I plan to restructure "my" API a bit anyway and will keep this in mind.

Interesting point about C++20 and `constexpr string`. Just as I was learning to not just pass by const reference, but thinking about move, views etc. I have to consider whether a const reference is actually the right thing to do just like it was pre-C++11. C++ is really making things simpler by making them more complicated :-)